### PR TITLE
[rls] Explicit usm-host memory copy routines

### PR DESCRIPTION
**Pull request by michael-smirnov**
Nightly: http://intel-ci.intel.com/ebcf4cba-fa1d-f1b3-aa94-8cdcd4b723a1

### CODE REVIEW

File: cpp/oneapi/dal/backend/memory.hpp [hunk 1]
```diff
@@ -112,23 +112,24 @@ inline T* malloc_host(const sycl::queue& queue, std::int64_t count) {
...
     ONEDAL_ASSERT(size > 0);
+    ONEDAL_ASSERT(is_known_usm(queue, dest));
+    ONEDAL_ASSERT(is_known_usm(queue, src));
     return queue.submit([&](sycl::handler& cgh) {
         cgh.depends_on(deps);
         cgh.memcpy(dest, src, size);
     });
 }
 
+inline sycl::event memcpy(sycl::queue& queue, void* dest, const void* src, std::size_t size) {
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/memory.hpp:129]: Why not to add default value to `const event_vector& deps` in the function above?

File: cpp/oneapi/dal/backend/memory.hpp [hunk 2]
```diff
@@ -149,6 +150,40 @@ inline sycl::event copy(sycl::queue& queue,
...
+                                 T* dest_usm,
+                                 const T* src_host,
+                                 std::int64_t count,
+                                 const event_vector& deps = {}) {
+    ONEDAL_ASSERT(count > 0);
+    ONEDAL_ASSERT(is_known_usm(queue, dest_usm));
+
+    // TODO: Remove additional copy to host usm memory once
+    //       bug in `copy` with the host memory is fixed
+    T* tmp_usm_host = malloc_host<T>(queue, count);
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/memory.hpp:164]: What about memory deallocation? Tip: use `unique_usm_ptr` defined below

File: cpp/oneapi/dal/detail/memory_impl_dpc.hpp [hunk 3]
```diff
@@ -33,6 +33,14 @@ ONEDAL_EXPORT void memcpy(const data_parallel_policy& policy,
...
                           const void* src,
                           std::int64_t size);
+ONEDAL_EXPORT void memcpy_usm2host(const data_parallel_policy& policy,
+                                   void* dest_host,
+                                   const void* src_usm,
+                                   std::int64_t size);
+ONEDAL_EXPORT void memcpy_host2usm(const data_parallel_policy& policy,
+                                   void* dest_usm,
+                                   const void* src_host,
+                                   std::int64_t size);
```
Comment (rlnx) [inline cpp/oneapi/dal/detail/memory_impl_dpc.hpp:43]: I do not see usage of those functions. Do we have `memcpy` in the header files that should be replaced by one those functions?

File: cpp/oneapi/dal/backend/memory.hpp (hunk 2 above)
Comment (michael-smirnov) [reply cpp/oneapi/dal/backend/memory.hpp:164]: Obvious memory leak 😢 
thank you!

File: cpp/oneapi/dal/detail/memory_impl_dpc.hpp (hunk 3 above)
Comment (michael-smirnov) [reply cpp/oneapi/dal/detail/memory_impl_dpc.hpp:43]: These functions are exposed in public API to use it internally in benchmarks

File: cpp/oneapi/dal/backend/memory.hpp [hunk 4]
```diff
@@ -112,23 +112,52 @@ inline T* malloc_host(const sycl::queue& queue, std::int64_t count) {
...
+                                   std::int64_t size,
+                                   const event_vector& deps = {}) {
+    ONEDAL_ASSERT(is_known_usm(queue, dest_usm));
+
+    // TODO: Remove additional copy to host usm memory once
+    //       bug in `copy` with the host memory is fixed
+    void* tmp_usm_host = malloc_host(queue, size);
+    memcpy(tmp_usm_host, src_host, size);
+    memcpy(queue, dest_usm, tmp_usm_host, size, deps).wait_and_throw();
+    free(queue, tmp_usm_host);
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/memory.hpp:141]: What if exception is thrown in memcpy functions?

File: cpp/oneapi/dal/backend/memory.hpp [hunk 5]
```diff
@@ -112,23 +112,52 @@ inline T* malloc_host(const sycl::queue& queue, std::int64_t count) {
...
+                                   std::int64_t size,
+                                   const event_vector& deps = {}) {
+    ONEDAL_ASSERT(is_known_usm(queue, src_usm));
+
+    // TODO: Remove additional copy to host usm memory once
+    //       bug in `copy` with the host memory is fixed
+    void* tmp_usm_host = malloc_host(queue, size);
+    memcpy(queue, tmp_usm_host, src_usm, size, deps).wait_and_throw();
+    memcpy(dest_host, tmp_usm_host, size);
+    free(queue, tmp_usm_host);
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/memory.hpp:157]: What if exception is thrown in memcpy functions?

**Comment by michael-smirnov** [comment]
@Mergifyio backport master

File: cpp/oneapi/dal/backend/memory.hpp [hunk 6]
```diff
@@ -213,6 +176,94 @@ inline unique_usm_ptr<T> make_unique_usm_host(const sycl::queue& q, std::int64_t
     return unique_usm_ptr<T>{ malloc_host<T>(q, count), usm_deleter<T>{ q } };
 }
 
+inline sycl::event memcpy(sycl::queue& queue,
+                          void* dest,
+                          const void* src,
+                          std::size_t size,
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/memory.hpp:182]: Here is `size_t`, but in other function like `memcpy_host2usm`, `memcpy_usm2host` `int64_t` is used

File: cpp/oneapi/dal/backend/memory.hpp [hunk 7]
```diff
@@ -213,6 +176,94 @@ inline unique_usm_ptr<T> make_unique_usm_host(const sycl::queue& q, std::int64_t
...
+    return {};
+}
+
+template <typename T>
+inline sycl::event copy(sycl::queue& queue, T* dest, const T* src, std::int64_t count) {
+    ONEDAL_ASSERT(count > 0);
+    const std::size_t n = detail::integral_cast<std::size_t>(count);
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::size_t, sizeof(T), n);
+    return memcpy(queue, dest, src, sizeof(T) * n);
+}
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/memory.hpp:229]: Remove this

File: cpp/oneapi/dal/backend/memory.hpp [hunk 8]
```diff
@@ -213,6 +176,94 @@ inline unique_usm_ptr<T> make_unique_usm_host(const sycl::queue& q, std::int64_t
...
+
+template <typename T>
+inline sycl::event copy(sycl::queue& queue,
+                        T* dest,
+                        const T* src,
+                        std::int64_t count,
+                        const event_vector& deps) {
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/memory.hpp:236]:
```suggestion
                        const event_vector& deps = {}) {
```

File: cpp/oneapi/dal/test/engine/dataframe_common.cpp [hunk 9]
```diff
@@ -256,6 +256,22 @@ static homogen_table wrap_to_homogen_table(device_test_policy& policy,
...
+                                     device_data,
+                                     data.get_data(),
+                                     data.get_size());
+        auto deleter = dal::detail::make_default_delete<const Float>(policy.get_queue());
+        return dal::homogen_table{ policy.get_queue(),
+                                   device_data,
+                                   row_count,
+                                   column_count,
+                                   deleter };
+    }
```
Comment (rlnx) [inline cpp/oneapi/dal/test/engine/dataframe_common.cpp:273]: I think, we should change this in `homogen_table_builder::copy_data`

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 10]
```diff
@@ -81,12 +82,37 @@ class homogen_table_builder_impl
...
+        ONEDAL_ASSERT(input_element_count * dtype_size == data.get_size());
+
+#ifdef ONEDAL_DATA_PARALLEL
+        auto input_q = data.get_queue();
+        if (input_q) {
+            return copy_data(*input_q, data.get_data(), input_element_count, 1);
+        }
+#endif
+        copy_data(data.get_data(), input_element_count, 1);
+    }
```
Comment (rlnx) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:114]: Use `__ONEDAL_IF_QUEUE__` and `__ONEDAL_IF_NO_QUEUE__`, see `dal/detail/common.hpp`, usage example 
https://github.com/oneapi-src/oneDAL/blob/80e130dae144cdc378acef095839bd0f92c742a9/cpp/oneapi/dal/test/engine/linalg/matrix.hpp#L525

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 11]
```diff
@@ -81,12 +82,37 @@ class homogen_table_builder_impl
 
     void copy_data(const void* data, std::int64_t row_count, std::int64_t column_count) override {
         check_copy_data_preconditions(row_count, column_count);
+
+#ifdef ONEDAL_DATA_PARALLEL
+        auto this_q = data_.get_queue();
+        if (this_q) {
+            ONEDAL_ASSERT(is_known_usm(data_));
+            return detail::memcpy_host2usm(*this_q,
```
Comment (rlnx) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:90]: Recommend to use `backend::memcpy_host2usm`

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 12]
```diff
@@ -81,12 +82,37 @@ class homogen_table_builder_impl
 
     void copy_data(const void* data, std::int64_t row_count, std::int64_t column_count) override {
         check_copy_data_preconditions(row_count, column_count);
+
+#ifdef ONEDAL_DATA_PARALLEL
+        auto this_q = data_.get_queue();
```
Comment (rlnx) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:87]: Use `__ONEDAL_IF_QUEUE__ `

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp [hunk 13]
```diff
@@ -121,11 +147,20 @@ class homogen_table_builder_impl
...
                    std::int64_t row_count,
                    std::int64_t column_count) override {
-        auto& queue = policy.get_queue();
-        ONEDAL_ASSERT(sycl::get_pointer_type(data_.get_data(), queue.get_context()) !=
-                      sycl::usm::alloc::unknown);
         check_copy_data_preconditions(row_count, column_count);
-        detail::memcpy(queue, data_.get_mutable_data(), data, data_.get_size());
+
+        auto& input_q = policy.get_queue();
+        auto this_q = data_.get_queue();
```
Comment (rlnx) [inline cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp:153]: You need to check if both queues are in the same context, see 
https://github.com/oneapi-src/oneDAL/blob/80e130dae144cdc378acef095839bd0f92c742a9/cpp/oneapi/dal/backend/common.hpp#L282

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp (hunk 11 above)
Comment (michael-smirnov) [reply cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp]: discussed offline

File: cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp (hunk 13 above)
Comment (michael-smirnov) [reply cpp/oneapi/dal/table/backend/homogen_table_builder_impl.hpp]: discussed offline

**Review by rlnx — APPROVED** [review]

**Pull request merged**